### PR TITLE
Merge Dev into main after beta release

### PR DESCRIPTION
**Pull request by okuchaiev**

### CODE REVIEW

**Comment by blisc** [comment]
Let's merge this?

**Comment by okuchaiev** [comment]
@blisc  just resolved few merge conflicts with main. removing WIP status

**Review by tkornuta-nvidia — APPROVED** [review]

File: nemo/utils/app_state.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nemo.utils.metaclasses import Singleton
+
+
+class AppState(metaclass=Singleton):
```
Comment (blisc) [inline nemo/utils/app_state.py:18]: What is this used for?

**Review by blisc — APPROVED** [review]
Good to go from the TTS side

File: nemo/utils/app_state.py (hunk 1 above)
Comment (ericharper) [reply nemo/utils/app_state.py:18]: It's used for Model Parallelism. For example, MegatronBertEncoder is a neural module, but needs world state information not normally available to neural modules.

Comment (ericharper) [reply nemo/utils/app_state.py:18]: Also, this is @tkornuta-nvidia implementation from old NeMo, I just copy-pasted it and added the properties I needed.

Comment (tkornuta-nvidia) [reply nemo/utils/app_state.py:18]: dozens of thumbs up for this :)

**Pull request merged**